Feature/acceptance testing

Tekkon opened the pull request:
Реализованы истории:

- Пользователь может создавать вопрос
- Пользователь может просматривать список вопросов
- Пользователь, находясь на странице вопроса, может написать ответ на вопрос (т.е. форма нового ответа должна быть прямо на странице вопроса, без перехода на другую страницу)
- Пользователь может просматривать вопрос и ответы к нему.
- Пользователь может войти в систему
- Пользователь может выйти из системы
- Пользователь может зарегистрироваться в системе
- Только аутентифицированный пользователь может создавать вопросы и ответы
- Автор может удалить свой вопрос или ответ, но не может удалить чужой вопрос/ответ

Commit:
Acceptance-тесты для создания вопроса и логина пользователя

Commit:
Пользователь может выйти из системы

Commit:
Пользователь может зарегистрироваться в системе

Commit:
- Пользователь может просматривать список вопросов
- Пользователь может просматривать вопрос и ответы к нему

Commit:
Пользователь, находясь на странице вопроса, может написать ответ на вопрос (т.е. форма нового ответа должна быть прямо на странице вопроса, без перехода на другую страницу)

Commit:
Только аутентифицированный пользователь может создавать вопросы и ответы

Commit:
Добавил автора к вопросу и ответу

Commit:
Автор может удалить свой вопрос или ответ, но не может удалить чужой вопрос/ответ

Commit:
Исправил спеки, которые упали после изменений в БД

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -1,4 +1,5 @@
+  before_action :authenticate_user!
   before_action :find_question, only: [:new, :create]
 
   def new
```
Нам не нужен экшн new, т.к. форма должна располагаться на странице вопроса

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -9,6 +10,7 @@ def create
+      flash[:notice] = 'Your answer is created successfully.'
       redirect_to question_path(@question)
     else
       render :new
```
Нужно рендерить страницу вопроса, ведь именно там расположена форма

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -18,7 +20,7 @@ def create
 
   def answer_params
-    params.require(:answer).permit(:body)
+    params.require(:answer).permit(:body, :user_id)
```
user_id не должен передаваться в параметрах, нужно связывать в контроллере с `current_user`

vkurennov commented on app/controllers/questions_controller.rb:
```diff
@@ -45,6 +53,6 @@ def load_question
 
   def question_params
-    params.require(:question).permit(:title, :body)
+    params.require(:question).permit(:title, :body, :user_id)
```
user_id не должен передаваться в параметрах, нужно связывать в контроллере с current_user

vkurennov commented on app/controllers/questions_controller.rb:
```diff
@@ -34,7 +36,13 @@ def update
 
   def destroy
-    @question.destroy
+    if @question.user_id == current_user.id
```
Такая проверка будет встречаться часто, вынеси ее в метод `User#author_of?` и используй везде, где нужно

vkurennov commented on app/views/layouts/application.html.erb:
```diff
@@ -3,12 +3,16 @@
+    <p class="alert"><%= alert %></p>
+
+    <% if user_signed_in? %>
+      <a href="/logout">Sign out</a>
```
А где ссылки на вход?

vkurennov replied:
Используй link_to, т.к. для выхода нужно еще и метод delete указать

vkurennov commented on app/views/questions/index.html.slim:
```diff
@@ -0,0 +1,9 @@
+- @questions.each do |q|
+  =p q.title
+  =p q.body
+  =p link_to 'Delete this question', question_path(q), method: :delete
```
Зачем показывать ссылку всем? Показывай только автору

vkurennov commented on app/views/questions/new.html.slim:
```diff
@@ -0,0 +1,7 @@
+= form_for @question do |f|
```
Здесь нужен вывод ошибок валидации

vkurennov commented on app/views/questions/new.html.slim:
```diff
@@ -0,0 +1,7 @@
+  = f.text_field :title
+  = f.label :body
+  = f.text_area :body
+  = f.hidden_field :user_id, value: user_signed_in? ? current_user.id : nil
```
Никогда так не делай. Во-первых, тут легко подставить любое значение и создать вопрос от имени другого пользователя, т.е. это дыра в безопасности. Во-вторых, current_user доступен и в контроллере. просто назначть атрибуту user значение из current_user и все.

vkurennov commented on app/views/questions/show.html.slim:
```diff
@@ -0,0 +1,17 @@
+
+h2 Answers
+- @question.answers.each do |a|
+  =p a.body
```
А как удалять ответы?

vkurennov commented on app/views/questions/show.html.slim:
```diff
@@ -0,0 +1,17 @@
+  =p a.body
+
+h2 Add an answer
+= form_for [@question, @question.answers.build] do |f|
```
Тут нужен вывод ошибок валидации

vkurennov commented on app/views/questions/show.html.slim:
```diff
@@ -0,0 +1,17 @@
+  p
+    = f.label :body
+    = f.text_field :body
+    = f.hidden_field :user_id, value: user_signed_in? ? current_user.id : nil
```
Связывай в контроллере, а тут убери эту строку

vkurennov commented on config/routes.rb:
```diff
@@ -1,4 +1,11 @@
+  root to: "questions#index"
+
+  devise_scope :user do
+    get 'logout', to: 'devise/sessions#destroy'
```
Нет, это костыль. Используй нормальную ссылку с method: :delete в видах

vkurennov commented on spec/controllers/answers_controller_spec.rb:
```diff
@@ -16,24 +18,26 @@
     context 'with valid attributes' do
       it 'saves the new answer' do
-        expect { post :create, params: { question_id: question, answer: attributes_for(:answer) } }.to change(question.answers, :count).by(1)
+        expect { post :create, params: { question_id: question, answer: attributes_for(:answer, user_id: user) } }.to change(question.answers, :count).by(1)
```
`user_id` не должен передаваться в параметрах

vkurennov commented on spec/controllers/answers_controller_spec.rb:
```diff
@@ -16,24 +18,26 @@
   describe 'POST #create' do
+    sign_in_user
+
     context 'with valid attributes' do
```
Нужен тест, проверяющий связь ответа и залогиненного юзера

vkurennov commented on spec/controllers/answers_controller_spec.rb:
```diff
@@ -16,24 +18,26 @@
 
       it 'redirects to question show view' do
-        post :create, params: { question_id: question, answer: attributes_for(:answer) }
+        post :create, params: { question_id: question, answer: attributes_for(:answer, user_id: user) }
```
`user_id` не должен передаваться в параметрах

vkurennov commented on spec/controllers/answers_controller_spec.rb:
```diff
@@ -16,24 +18,26 @@
     context 'with invalid attributes' do
       it 'does not save the question' do
-        expect { post :create, params: { question_id: question, answer: attributes_for(:invalid_answer) } }.to_not change(Answer, :count)
+        expect { post :create, params: { question_id: question, answer: attributes_for(:invalid_answer, user_id: user) } }.to_not change(Answer, :count)
```
`user_id` не должен передаваться в параметрах

vkurennov commented on spec/controllers/answers_controller_spec.rb:
```diff
@@ -1,9 +1,11 @@
 require 'rails_helper'
 
 RSpec.describe AnswersController, type: :controller do
```
Где тесты на метод destroy? Прежде чем писать какой-то код, остановись и подумай, есть ли у тебя тесты на него? Если нет, то начни с тестов, продумай, как метод должен себя вести и напиши тесты на это шаг за шагом.

vkurennov commented on spec/controllers/answers_controller_spec.rb:
```diff
@@ -16,24 +18,26 @@
 
       it 're-renders new view' do
-        post :create, params: { question_id: question, answer: attributes_for(:invalid_answer) }
+        post :create, params: { question_id: question, answer: attributes_for(:invalid_answer, user_id: user) }
```
`user_id` не должен передаваться в параметрах

vkurennov commented on spec/controllers/questions_controller_spec.rb:
```diff
@@ -1,10 +1,11 @@
 
   describe 'GET #index' do
-    let(:questions) { create_list(:question, 2) }
+    let(:questions) { [create(:question, user: user), create(:question, user: user)] }
```
Используй create_list, он тоже умеет принимать доп. параметры после указания кол-ва

vkurennov commented on spec/controllers/questions_controller_spec.rb:
```diff
@@ -54,30 +57,34 @@
   describe 'POST #create' do
+    sign_in_user
+
     context 'with valid attributes' do
```
Нужен тест на связь вопроса и залогиненного юзера

vkurennov commented on spec/controllers/questions_controller_spec.rb:
```diff
@@ -54,30 +57,34 @@
     context 'with valid attributes' do
       it 'saves the new question in the database' do
-        expect { post :create, params: { question: attributes_for(:question) } }.to change(Question, :count).by(1)
+        expect { post :create, params: { question: attributes_for(:question, user_id: user) } }.to change(Question, :count).by(1)
```
`user_id` не должен передаваться в параметрах

vkurennov commented on spec/controllers/questions_controller_spec.rb:
```diff
@@ -54,30 +57,34 @@
 
       it 'redirects to show view' do
-        post :create, params: { question: attributes_for(:question) }
+        post :create, params: { question: attributes_for(:question, user_id: user) }
```
`user_id` не должен передаваться в параметрах

vkurennov commented on spec/controllers/questions_controller_spec.rb:
```diff
@@ -54,30 +57,34 @@
     context 'with invalid attributes' do
       it 'does not save the question' do
-        expect { post :create, params: { question: attributes_for(:invalid_question) } }.to_not change(Question, :count)
+        expect { post :create, params: { question: attributes_for(:invalid_question, user_id: user) } }.to_not change(Question, :count)
```
`user_id` не должен передаваться в параметрах

vkurennov commented on spec/controllers/questions_controller_spec.rb:
```diff
@@ -54,30 +57,34 @@
 
       it 're-renders new view' do
-        post :create, params: { question: attributes_for(:invalid_question) }
+        post :create, params: { question: attributes_for(:invalid_question, user_id: user) }
```
`user_id` не должен передаваться в параметрах

vkurennov commented on spec/controllers/questions_controller_spec.rb:
```diff
@@ -116,10 +123,12 @@
     before { question }
 
```
У тебя в удалении есть условие, значит, надо тестировать все варианты. Сделай 2 контекста: для автора и для не-автора

vkurennov commented on spec/features/browse_questions_spec.rb:
```diff
@@ -0,0 +1,39 @@
+
+  given(:question) { create(:question, user: create(:user)) }
+
+  scenario 'User browses a list of questions' do
```
Если тестируешь список вопросов, то и проверяй **список**, а не один вопрос. Причем вопросы должны быть разные (используй sequence в фабриках для этого)

vkurennov commented on spec/features/browse_questions_spec.rb:
```diff
@@ -0,0 +1,39 @@
+  given(:question) { create(:question, user: user) }
+
+  scenario 'User browses a question and its answers' do
+    answer = create(:answer, question: question, user: user)
```
И опять же проверяй **список ответов**, а не один ответ. И также все ответы должны быть разными

vkurennov commented on spec/features/create_answer_spec.rb:
```diff
@@ -0,0 +1,32 @@
+    fill_in 'Body', with: 'This is my answer.'
+    click_on 'Create an answer'
+
+    expect(page).to have_content question.answers.last.body
```
Лучше проверить непосредственно текст, который указан в 16-й строке

vkurennov commented on spec/features/create_answer_spec.rb:
```diff
@@ -0,0 +1,32 @@
+require 'rails_helper'
+
+feature 'User can create an answer on the question page', %q{
```
Нужен тест на невалидный ответ - проверить, что система сообщает об ошибках валидации

vkurennov commented on spec/features/create_question_spec.rb:
```diff
@@ -0,0 +1,30 @@
+require 'rails_helper'
+
+feature 'Create question', %q{
```
Нужен тест на невалидный вопрос

vkurennov commented on spec/features/create_question_spec.rb:
```diff
@@ -0,0 +1,30 @@
+    fill_in 'Body', with: 'test text'
+    click_on 'Create'
+
+    expect(page).to have_content 'Your question is created successfully.'
```
Проверь наличие заголовка и тела вопроса

vkurennov commented on spec/features/delete_answer_spec.rb:
```diff
@@ -0,0 +1,37 @@
+    create(:question, user_id: user.id)
+    visit questions_path
+    click_on 'Delete this question'
+    expect(page).to have_content 'Your question is deleted successfully.'
```
Надо проверить, что текста вопроса больше нет на странице

vkurennov commented on spec/features/delete_answer_spec.rb:
```diff
@@ -0,0 +1,37 @@
+  scenario 'Authenticated user tries to delete his question' do
+    sign_in(user)
+
+    create(:question, user_id: user.id)
```
Лучше такие вещи выносить в let

vkurennov commented on spec/features/delete_answer_spec.rb:
```diff
@@ -0,0 +1,37 @@
+    another_user = create(:user)
+    create(:question, user_id: another_user.id)
+    visit questions_path
+    click_on 'Delete this question'
```
Лучше вообще не показывать ссылку на удаление и проверить, что этой ссылки нет на странице

vkurennov reviewed (changes requested):
Кроме прочего, нет сценария удаления ответа.

Commit:
- Убрал экшн new, т.к. форма должна располагаться на странице вопроса
- В create рендерим страницу вопроса, т.к. именно там расположена форма

Commit:
Убрал передачу user_id в параметрах, сделал связку с current_user в контроллерах

Commit:
- Добавил метод author_of? в модель User для проверки, является ли пользователь автором вопроса/ответа
- Переименовал спек удаления вопроса, по ошибке он был назван неправильно: delete_answer_spec

Commit:
Убрал скрытые поля, в которые подставлялся user_id

Commit:
Ссылку на удаление вопроса сделал видимой только для автора

Commit:
Добавил тесты на удаление вопроса, кнопку для удаления вопроса, реализовал метод destroy в AnswersController

Commit:
- Создал секвенции в фабриках для вопроса и ответа
- В тестах на проверку списков вопросов и ответов добавил создание списков из секвенции фабрики
- Expect'ы переписал на проверку наличия каждого объекта из списка

Commit:
- Добавил паршал для вывода ошибок валидации
- В спеки создания вопроса и ответа добавил тесты на проверку вывода ошибок валидации

Commit:
Убрал экшен new из AnswerController, т.к. новый ответ создается на странице вопроса.

Commit:
- Убрал route на logout
- Доавил ссылки Sign in, Sign out, Sign up в layout
- Переписал тесты входа, выхода и регистрации с использованием этих ссылок

Commit:
- Добавил тест, проверяющий связь вопроса и залогиненного юзера
- Добавил тест, проверяющий связь ответа и залогиненного юзера

Commit:
Разделил тесты удаления вопроcа/ответа на два контекста: 1) когда автор удаляет вопрос/ответ и 2) когда не автор удаляет вопрос/ответ

Commit:
- Перенес создание вопроса/ответа в given
- Добавил проверку, что текста вопроса/ответа нет на странице после удаления

Commit:
Изменид присвоение user_id на присовение user

Commit:
Поправил тест, который упал после изменений в фабрике

Commit:
В create контроллера AnswerController вместо редиректа добавил рендер страницы вопроса и переписал несколько тестов, которых это затронуло

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -1,18 +1,27 @@
-      redirect_to question_path(@question)
+    flash[:notice] = 'Your answer is created successfully.' if @answer.save
+
+    render 'questions/show'
```
Почему рендер безусловный? В случае успешного сохранения, нужно делать редирект на вопрос, тогда подгрузятся другие ответы, если они в это время были даны и форма очистится, а иначе, форма создания ответа автоматом превратится в форму редактирования ответа

vkurennov commented on app/views/questions/show.html.slim:
```diff
@@ -0,0 +1,19 @@
+- @question.answers.each do |a|
+  =p a.body
+  - if user_signed_in? && current_user.author_of?(a)
+    =p link_to('Delete this answer', question_answer_path(@question,a), method: :delete)
```
У link_to и прочих view-хелперов не принято ставить скобки у аргументов.

vkurennov commented on app/views/questions/show.html.slim:
```diff
@@ -0,0 +1,19 @@
+
+h2 Add an answer
+= form_for [@question, @answer] do |f|
+  = render "shared/error_messages", :target => @answer
```
Используй всегда новую форму записи хешей: `target: @answer`

vkurennov commented on config/routes.rb:
```diff
@@ -1,4 +1,7 @@
+  root to: "questions#index"
+
   resources :questions do
   	resources :answers
```
Что с отступами?

vkurennov commented on spec/controllers/answers_controller_spec.rb:
```diff
@@ -1,29 +1,27 @@
+      it 'sets a current_user to the new answer' do
+        sign_in_the_user(user)
+        post :create, params: { question_id: question, answer: attributes_for(:answer) }
+        expect(question.answers.last.user_id).to eq user.id
```
Если в какой-то момент изменится дефолтная сортировка, тест будет нерелевантным. Не обязательно он упадет, но может быть ложно-положительным. Поэтому лучше взять ответ непосредственно из контроллера, используй метод `assign`

vkurennov commented on spec/controllers/questions_controller_spec.rb:
```diff
@@ -54,11 +57,19 @@
+      it 'sets a current_user to the new question' do
+        sign_in_the_user(user)
+        post :create, params: { question: attributes_for(:question) }
+        expect(question.user_id).to eq user.id
```
question - это не созданный в контроллере вопрос, это вопрос, созданный через фабрики. У него просто такой же user_id, поэтому тест ложно-положительный. Надо брать объект непосредственно из контроллера через `assign`

vkurennov commented on spec/factories/answers.rb:
```diff
@@ -1,8 +1,16 @@
     body "MyText"
   end
 
+  factory :sequence_answer, class: 'Answer' do
```
Зачем? можно было в основной фабрике примениь sequence и не делать еще одну

vkurennov commented on spec/factories/questions.rb:
```diff
@@ -1,7 +1,16 @@
+    body 'I really want to know!'
+  end
+
+  factory :sequence_question, class: 'Question' do
```
Также можно было в фабрике `question` сделать динамические заголовок и тело

vkurennov commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,6 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
```
Где тесты на метода `author_of?`

Commit:
- Если вопрос успешно сохранен, делаем редирект на страницу вопроса. Если вопрос не сохранен, делаем рендер страницы вопроса.
- Исправил тест проверки редиректа

Commit:
убрал скобки

Commit:
Использовал новую форму записи хеша

Commit:
Поправил отступы

Commit:
Редирект в случае успешного удаления и рендер в противном случае

Commit:
Замнил на assigns, чтобы проверять именно тот answer, который был создан в контроллере

Commit:
Заменил на assigns, чтобы проверять именно тот question, который был создан в контроллере

Commit:
Удалил лишние фабрики и переписал тесты, с этим связанные

Commit:
Добавил проверку на a.persisted?, т.к. при ошибке валидации вопрос не сохраняется в базе, поэтому у него нет id, в следствие чего возникала ошибка на этой строчке

Commit:
- Добавил тесты на ассоциации вопрос-юзер, ответ-юзер
- Добавил тесты на метод author_of? модели User

Commit:
Пропускаем вопросы, которые не были сохранены в БД, чтобы не выводить тело такого вопроса и ссылку на его удаление

vkurennov commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,37 @@
+      user.author_of?(another_question).should eq false
+    end
+
+    it 'returns true if user is the quthor of the answer' do
```
Не нужно на всех вариантах аргумента проверять, у тебя же метод универсальный. Если бы ты тестировал метод, который складывает 2 числа, ты бы ведь не стал писать все возможные варианты от 0 до бесконечности. Вот и тут также. Достатончо только на вопросе проверить

vkurennov commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,37 @@
+    let(:another_answer) { create(:answer, user: another_user, question: another_question) }
+
+    it 'returns true if user is the author of the question' do
+      user.author_of?(question).should eq true
```
* rspec делает динамические метчеры (методы для проверки) для методов-предикатов (это методы, которые заканчиваются на знак вопроса и возвращают true/false) и можно писать короче и красивее:
`expect(user).to be_author_of(question)`
т.е. название метчера получается из названия метода без вопр. знака и с доавлением `be_` в начале

Commit:
Использовал динамические метчеры
